### ISSUE TITLE
DataStore selective sync will still perform a scan when it should perform a query

### ISSUE THREAD

**Issue by l-hendriks** [aws-amplify/amplify-js#7923]
**Describe the bug**
When using selective sync to perform a query instead of a scan (explained here: https://docs.amplify.aws/lib/datastore/sync/q/platform/js#advanced-use-case---query-instead-of-scan) it will still perform a scan but uses the `syncExpression` as a filter. The model name is `Product` in this example, with `owner` as the key.

The schema looks like this:

```
type Product @model
  @auth(rules: [
    { allow: owner, ownerField: "owner" },
  ])
  @key(name: "byOwner", fields: ["owner", "id"], queryField: "getProductByOwner") {
  id: ID!
  owner: String
  ...
}
```

e.g. when I apply `syncExpression(Product, () => (item) => item.owner('eq', userId)),` in the DataStore.configure syncExpressions object, I will indeed get the filtered items (I have tested that the syncExpression works correctly using a second ownerField and creating a new record where the owner is in the second ownerField. Without the syncExpression I get both records, but with the syncExpression only the record with the owner field record set as the owner). However, when verifying the request type in the DynamoDB console, I still see scan results (see screenshot):

![image](https://user-images.githubusercontent.com/4309740/110770334-7fa2ad80-8259-11eb-85fe-a8ac4090ffc8.png)

I have tried to debug the syncProduct resolver, which has the following relevant part:

```
## [Start] Set map initialization for @key **
#set( $index = "" )
#set( $scan = true )
#set( $filterMap = {} )
#set( $QueryMap = {} )
#set( $PkMap = {} )
#set( $filterArgsMap = {} )
## [End] Set map initialization for @key **
## [Start] Set query expression for @key **
$util.qr($QueryMap.put('owner+id' , 'byOwner'))
$util.qr($PkMap.put('owner' , 'byOwner'))
## [End] Set query expression for @key **
## [Start] Set query expression for @key **
#set( $filterArgsMap = $ctx.args.filter.get("and") )
#if( !$util.isNullOrEmpty($filterArgsMap) && $util.isNull($ctx.args.lastSync) )
  #set( $json = $filterArgsMap )
  #foreach( $item in $json )
    #set( $ind = $foreach.index )
    #foreach( $entry in $item.entrySet() )
      #if( $ind == 0 && !$util.isNullOrEmpty($entry.value.eq) && !$util.isNullOrEmpty($PkMap.get($entry.key)) )
        #set( $pk = $entry.key )
        #set( $scan = false )
        $util.qr($ctx.args.put($pk,$entry.value.eq))
        #set( $index = $PkMap.get($pk) )
      #end
      #if( $ind == 1 && !$util.isNullOrEmpty($pk) && !$util.isNullOrEmpty($QueryMap.get("${pk}+$entry.key")) )
        #set( $sk = $entry.key )
        $util.qr($ctx.args.put($sk,$entry.value))
        #set( $index = $QueryMap.get("${pk}+$sk") )
      #else
        #if( $ind > 0 )
          $util.qr($filterMap.put($entry.key,$entry.value))
        #end
      #end
    #end
  #end
#else
  #set( $filterMap = $ctx.args.filter )
#end
```

By default "scan" is true (so it will perform a scan instead of a query), and will only be set to false if a couple of statements hold true.

Most importantly is `$util.isNull($ctx.args.lastSync)` in the first if statement. The lastSync argument needs to be null, but DataStore sends the value `0` there. This is the request from `DataStore.query()`:

```
{
  "query": "query operation($limit: Int, $nextToken: String, $lastSync: AWSTimestamp, $filter: ModelProductFilterInput) {\n  syncProducts(limit: $limit, nextToken: $nextToken, lastSync: $lastSync, filter: $filter) {\n    items {\n      id\n      owner\n      originalOwner\n      name\n      _version\n      _lastChangedAt\n      _deleted\n    }\n    nextToken\n    startedAt\n  }\n}\n",
  "variables": {
    "limit": 1000,
    "nextToken": null,
    "lastSync": 0,
    "filter": {
      "and": [
        {
          "owner": {
            "eq": "someId"
          }
        }
      ]
    }
  }
}
```

As you can see here, `lastSync` is equal to `0`, which is not `null`, so AppSync will always use a scan instead of a query.

I removed the `$util.isNull($ctx.args.lastSync)` part in the if statement and checked again. Now AppSync actually does a query:

![image](https://user-images.githubusercontent.com/4309740/110771540-e83e5a00-825a-11eb-8f8a-ccd4008cd208.png)

Either the appsync resolver should be updated to also use a query when `lastSync` is `0`, or DataStore should not send the `lastSync` parameter instead of a `0`.

**To Reproduce**

1. Use selective sync as described here: https://docs.amplify.aws/lib/datastore/sync/q/platform/js#advanced-use-case---query-instead-of-scan
2. Notice in the dynamodb metrics a scan is always performed

**Expected behavior**
A query should have been performed instead of a scan

**Code Snippet**
See above

**Screenshots**
See above

**Comment by l-hendriks** [comment]
Additional question: why is the lastSync parameter even in the if statement in the first place? If the lastSync was set before, it should still perform a query

**Comment by iartemiev** [comment]
Hey @l-hendriks, thanks for opening the issue and providing a detailed description.

The expected behavior is the following:
* When performing a Base Sync (`lastSync` === 0), attempt to construct a query expression from the selective sync predicate and perform a DDB Query against the base table if possible.

* If we're inside the full sync interval (24 hours by default, but configurable), perform a Delta Sync, which retrieves values from the Delta table that have been changed since the `lastSync` timestamp.

In other words, the selective sync predicate should only affect the Base Sync. The Delta Sync is already an optimized retrieval mechanism.

I think the solution should be updating the conditional in the resolver to:
```
#if( !$util.isNullOrEmpty($filterArgsMap) && ($util.isNull($ctx.args.lastSync) || $ctx.args.lastSync == 0))
```

I'll open a PR in the CLI repo that makes this change in the generated resolvers.

**Comment by l-hendriks** [comment]
Hi @iartemiev, thanks for the quick response. While writing the question in my comment I actually found out indeed that the Delta sync is using a different table altogether. So I think that will do the trick indeed.

Thanks again and have a nice day!

**Comment by l-hendriks** [comment]
I just tested this with 4.46.0 and a forced redeploy of the api and can confirm it now does a query instead of a scan. Thanks a lot!
